File: services/graph/graph.py
from __future__ import annotations

from typing import Literal
from dataclasses import dataclass, field

from langgraph.graph import StateGraph, START, END

from services.graph.nodes import AgentNodeRunner
from services.graph.state import AgentState


Route = Literal["tools", "__end__"]


@dataclass
class AgentGraphFactory:
    nodes: AgentNodeRunner = field(default_factory=AgentNodeRunner)

    def route_after_agent(self, state: AgentState) -> Route:
        last = state["messages"][-1]
        if last.get("tool_calls"):
            return "tools"
        return "__end__"

    def build(self):

        builder = StateGraph(AgentState)
        builder.add_node("agent", self.nodes.call_model)
        builder.add_node("tools", self.nodes.execute_tools)
        builder.add_edge(START, "agent")
        builder.add_conditional_edges(
            "agent",
            self.route_after_agent,
            {
                "tools": "tools",
                "__end__": END,
            },
        )
        builder.add_edge("tools", "agent")
        return builder.compile()

File: apps/server/src/graph/graph.ts
import { END, START, StateGraph } from "@langchain/langgraph";

import { getCheckpointer } from "../checkpoint.js";
import { callModel, executeTools, routeAfterAgent } from "./nodes.js";
import { AgentStateAnnotation } from "./state.js";

export function buildGraph(options?: { checkpointer?: boolean }) {
  const builder = new StateGraph(AgentStateAnnotation)
    .addNode("agent", callModel)
    .addNode("tools", executeTools)
    .addEdge(START, "agent")
    .addConditionalEdges("agent", routeAfterAgent, {
      tools: "tools",
      __end__: END,
    })
    .addEdge("tools", "agent");

  if (options?.checkpointer === false) {
    return builder.compile();
  }
  return builder.compile({ checkpointer: getCheckpointer() });
}
